autoparallel/activation_checkpointing.py

```diff
@@ -66,6 +66,91 @@ def is_wait_tensor_from_fsdp(node: torch.fx.Node) -> bool:
+            user = next(iter(node.users))
+            if len(user.all_input_nodes) > 1:
+                break
+            node = user
+            if is_all_gather_into_tensor(node):
+                last_ag_chain_node = node
+        if last_ag_chain_node is not None:
+            # 2. Find last wait_tensor from last all_gather
+            last_ag_wait_node = next(iter(last_ag_chain_node.users))
+            assert is_wait_tensor(last_ag_wait_node)
+            assert is_wait_tensor_from_fsdp(last_ag_wait_node)
+            # 3. Continue the linear chain from the last wait_tensor
+            w = last_ag_wait_node
+            while True:
+                if len(w.users) != 1:
+                    # Capture this pattern:
+                    # %wait_tensor_5 : [num_users=1] = call_function[target=torch.ops._c10d_functional.wait_tensor.default](args = (%all_gather_into_tensor_5,), kwargs = {})  # noqa: E501
+                    # %split : [num_users=4] = call_function[target=torch.ops.aten.split.Tensor](args = (%wait_tensor_5, 576), kwargs = {})
+                    # %getitem_2 : [num_users=1] = call_function[target=operator.getitem](args = (%split, 0), kwargs = {})
+                    # %getitem_3 : [num_users=1] = call_function[target=operator.getitem](args = (%split, 1), kwargs = {})
+                    # %getitem_4 : [num_users=1] = call_function[target=operator.getitem](args = (%split, 2), kwargs = {})
+                    # %getitem_5 : [num_users=1] = call_function[target=operator.getitem](args = (%split, 3), kwargs = {})
+                    # %cat_1 : [num_users=1] = call_function[target=torch.ops.aten.cat.default](args = ([%getitem_2, %getitem_3, %getitem_4, %getitem_5], 1), kwargs = {})  # noqa: E501
+                    if (
+                        w.op == "call_function"
+                        and w.target == torch.ops.aten.split.Tensor
+                    ):
+                        if all(
+                            split_user.op == "call_function"
+                            and split_user.target == operator.getitem
+                            and len(split_user.users) == 1
+                            for split_user in w.users
+                        ):
+                            getitem_users = list(
+                                next(iter(getitem_node.users))
+                                for getitem_node in w.users
+                            )
+                            potential_cat_op = getitem_users[0]
+                            if all(
+                                potential_cat_op == getitem_user
+                                for getitem_user in getitem_users
+                            ) and (
+                                potential_cat_op.target == torch.ops.aten.cat.default
+                            ):
+                                w = potential_cat_op
+                                continue
+                    break
+                user = next(iter(w.users))
```

Comment:
just to make sure i'm following, (maybe comment would help) - this would be the user of (a) the cat, if split-cat found or (b) the wait op?

Reply:
oh- no- i think what I said above is true on the first iteration of this line, but then subsequent iterations would keep advancing as long as its a linear chain, then when we break out this should be 'the first node in the usage chain with multiple users'

Reply:
For your first question, user = cat's next op only if cat op has one user. The second statement is correct.